contracts/src/Teleporter/upgrades/TeleporterUpgradeable.sol

```diff
@@ -0,0 +1,61 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "./TeleporterRegistry.sol";
+
+/**
+ * @dev TeleporterUpgradeable provides upgrade utility for applications built on top
+ * of the Teleporter protocol by integrating with the {TeleporterRegistry}.
+ */
+abstract contract TeleporterUpgradeable {
+    TeleporterRegistry public immutable teleporterRegistry;
+    uint256 internal _minTeleporterVersion;
+
+    error InvalidTeleporterRegistryAddress();
+    error InvalidTeleporterSender();
+
+    /**
+     * @dev Throws if called by a `msg.sender` that is not an allowed Telepoter version.
+     * Checks that `msg.sender` matches a Teleporter version greater than or equal to `_minTeleporterVersion`.
+     */
+    modifier onlyAllowedTeleporter() {
+        if (
+            teleporterRegistry.getAddressToVersion(msg.sender) <
+            _minTeleporterVersion
+        ) {
+            revert InvalidTeleporterSender();
+        }
+        _;
+    }
+
+    /**
+     * @dev Initializes the {TeleporterUpgradeable} contract by getting `teleporterRegistry`
+     * instance and setting `_minTeleporterVersion`.
+     */
+    constructor(address teleporterRegistryAddress) {
+        if (teleporterRegistryAddress == address(0)) {
+            revert InvalidTeleporterRegistryAddress();
+        }
+
+        teleporterRegistry = TeleporterRegistry(teleporterRegistryAddress);
+        _minTeleporterVersion = teleporterRegistry.getLatestVersion();
+    }
+
+    /**
+     * @dev Updates `_minTeleporterVersion` to the latest version.
+     */
+    function updateMinTeleporterversion() external {
```

Comment:
```suggestion
    function updateMinTeleporterVersion() external {
```

Reply:
Is `updateMinTeleporterVersion` a function that we want to restrict to "admin only" permission level? 

I think the expected use case is that a dApp would update the minimum Teleporter version once a newer version has been deployed and registered on both chains, and once they are confident that they don't need to receive any messages sent by either chain before it was updated to use the new version. 

Allowing anyone to call it introduces a time window where anyone can force that messages are only received from the latest version prior to the opposite chain being updated to send using the new version.

Reply:
I agree that it will likely be admin restricted for `updateMinTeleporterVersion` for most dapps, how were you thinking of going about it? There are so many variations of creating admins/governance, and not just by ownership, I was going to leave that up to dapps. Could also explore something like `TeleporterOwnable`.

Reply:
Good point, dApps could certainly have a bunch of different types of access control logic, and we don't want to pin them down here as far as having to define who an "admin" is. 

Is it possible to have the default be the contract deployer (using OpenZeppelin's [Ownable](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol)), but also make it easy for dApps to override the default behavior? Maybe something like a virtual `onlyUpdater` modifier.

Reply:
Looking at OZ contract's, each of these abstract contracts have a very specific purpose. I think we should leave `TeleporterUpgradeable` as a base level helper for integrating with the registry. Then for different types of access controls we'd have individual contracts, for example we can have one for `OwnerUpgradeable`, `AdminUpgradeable`, etc.

Reply:
Hm, yeah that makes sense, but in this case dApps that inherit directly from `TeleporterUpgradable` would be vulnerable to having time window where anyone can force that messages are only received from the latest version prior to the opposite chain being updated to send using the new version. I think we should prevent this case by default.